src/Aspire.Dashboard/Otlp/Model/OtlpTrace.cs

```diff
@@ -96,6 +101,31 @@ static string BuildFullName(OtlpSpan existingSpan)
         }
     }
 
+    private static bool HasCircularReference(OtlpSpan span)
+    {
+        // Can't have a circular reference if the span has no parent.
+        if (string.IsNullOrEmpty(span.ParentSpanId))
+        {
+            return false;
+        }
+
+        // Walk up span ancestors to check there is no loop.
+        var stack = new List<OtlpSpan> { span };
+        var currentSpan = span;
+        while (currentSpan.GetParentSpan() is { } parentSpan)
+        {
+            if (stack.Contains(parentSpan))
```

Comment:
Is a `List<>` the right collection here? Seems like a `HashSet<>` would be a better fit, avoiding O(N) lookups.

Concerned about the space and time complexity of this change. As called out above, the method's called for every span, and it then walks the whole ancestry, and does an O(N) lookup within that. That's a lot of nested loops.

Reply:
Changed to a keyed collection. Will adaptively switch to a dictionary when it exceeds a certain number of spans.